### PR TITLE
DocumentBar: replace icon with post type label

### LINKED ISSUE #65167: Update DocumentBar with post type label
In an effort to improve the reference points between post types in the Site Editor, I propose removing the icons and using a visible label for the post types in the DocumentBar. 

### Tasks

- [x] Remove the icon. 
- [x] Add support for the post type label. 
- [x] If there's a post type label, place a `·` between them.

### Current 

current

### Proposed 

proposed

### PR DESCRIPTION
**Pull request by creativecoder**
## What?

Replaces the post type icon in the document bar with an explicit label

Resolves #65167

## Why?

To help make it more obvious exactly what kind of document you are editing.

## How?

- Add post type label to DocumentBar for the currently edited document (except for Style Book and Global Style revisions, since those are currently rendered differently)
- Remove the default document icon from DocumentBar (but still display the icon prop if explicitly passed, for backward compatibility)

## Testing Instructions

- Open various kinds of post types in the editor (posts/page, templates, navigation menu, patterns, etc.
- Open Style Book and Global Style Revisions to see that there is no post type displayed

### Testing Instructions for Keyboard

n/a

## Screenshots or screencast 

| **Before** | **After** |
| - | - |
| image | image |

### CODE REVIEW

File: packages/editor/src/components/document-bar/index.js
Comment (creativecoder) [inline packages/editor/src/components/document-bar/index.js:200]: The aria-label doesn't seem necessary any more, since the post type is now displayed after the document title.

File: packages/editor/src/components/document-bar/index.js
Comment (creativecoder) [inline packages/editor/src/components/document-bar/index.js:58]: "import("@wordpress/components")" is currently output [in the documentation](https://github.com/WordPress/gutenberg/tree/trunk/packages/editor#documentbar), so I've changed the param type here to just `IconType`.

I'm not sure what a good way is to indicate the type comes from `@wordpress/components`.

File: packages/editor/src/components/document-bar/index.js
Comment (andrewserong) [reply packages/editor/src/components/document-bar/index.js:200]: I noticed that this removal causes some of the e2e tests to fail. E.g. this one: https://github.com/WordPress/gutenberg/blob/50f5c74ecbd267a701704d607f70385255a72033/test/e2e/specs/editor/various/post-editor-template-mode.spec.js#L40

The failure appears to be the check on this line: https://github.com/WordPress/gutenberg/blob/50f5c74ecbd267a701704d607f70385255a72033/test/e2e/specs/editor/various/post-editor-template-mode.spec.js#L198-L200

The aria-label does add a bit more context as it explains the context (editing) rather than using the visual separator. So I wonder if it's worth still including it? Either that, or updating the test 🙂

I'll just ping @jeryj on this one, too, as I remember there being previous discussions about the accessibility of this bar.

File: packages/editor/src/components/document-bar/index.js
Comment (andrewserong) [reply packages/editor/src/components/document-bar/index.js:58]: If we still want to indicate that it comes from `@wordpress/components` within the file, but not in the docs, could we try something like the following slightly further up the file?

```
/** @typedef {import("@wordpress/components").IconType IconType */
```

I don't know how legitimate that is, but just looking at a couple of examples elsewhere that appear to do that:

https://github.com/WordPress/gutenberg/blob/3ea2d42b0a6a206663735a47f9796bd42eda2186/packages/blocks/src/api/parser/serialize-raw-block.js#L11

and

https://github.com/WordPress/gutenberg/blob/c56672504bada434aee1ec331366cd2a10ab2b38/packages/block-editor/src/components/list-view/index.js#L69-L70

**Comment by andrewserong** [comment]
This is looking really nice to me in testing! Just pushed a tiny update to try to fix one of the e2es. Looks like there's another that will need fixing (or for us to restore the aria-label): https://github.com/WordPress/gutenberg/pull/65170#discussion_r1751131846

✅ Looking good in the page editor, post editor, and site editor:
image

✅ Looking good in custom post type editor:
image

❓ One thing I noticed while testing is that the DocumentBar appears to be displayed in bold in RTL languages (otherwise RTL appears to be working pretty well):

image

This appears to be because of rules output via the localization stylesheet:

image

To resolve that, I think we could add `font-weight: 400' to the `h1` rule here:

https://github.com/WordPress/gutenberg/blob/7ba8423a897212598d4f7442e1fd77c1cf0ddcd3/packages/editor/src/components/document-bar/style.scss#L59-L65

Does that sound worth doing in this PR? If not, happy to open a separate one-liner PR for it 🙂

File: packages/editor/src/components/document-bar/style.scss
Comment (jasmussen) [inline packages/editor/src/components/document-bar/style.scss]: This may need to remain $gray-800, or be #6d6d6d to meet 4.5:1 contrast. https://contrast.tools is a great way to test.

Other than that, ship it!

**Comment by jameskoster** [comment]
Screenshot 2024-09-10 at 10 05 40

Should we display the `area` here rather than the generic "Template part"?

File: packages/editor/src/components/document-bar/index.js
Comment (stokesman) [inline packages/editor/src/components/document-bar/index.js]: This might have gotten auto-inserted with an editor completion. It doesn’t have a purpose in a .js file as the type only appears in jsdoc.
```suggestion
```

**Review by stokesman — COMMENTED** [review]
I note that with long titles the type label will be truncated: 
image

Maybe it’d be preferable to keep the type label visible:
image

<details><summary>Diff to have title truncated separately from post type</summary>

```diff
diff --git a/packages/editor/src/components/document-bar/index.js b/packages/editor/src/components/document-bar/index.js
index be291d6be5..84b12a7d42 100644
--- a/packages/editor/src/components/document-bar/index.js
+++ b/packages/editor/src/components/document-bar/index.js
@@ -181,12 +181,14 @@ export default function DocumentBar( props ) {
 							{ title
 								? decodeEntities( title )
 								: __( 'No title' ) }
-
-							{ postTypeLabel && ! props.title && (
-								<span className="editor-document-bar__post-type-label">
-									{ ' · ' + decodeEntities( postTypeLabel ) }
-								</span>
-							) }
+						</Text>
+						<Text
+							size="body"
+							className="editor-document-bar__post-type-label"
+						>
+							{ postTypeLabel &&
+								! props.title &&
+								' · ' + decodeEntities( postTypeLabel ) }
 						</Text>
 					</motion.div>
 					<span className="editor-document-bar__shortcut">

```
</details>

Note that doing so would mean responsive styling will need some attention:
image

File: packages/editor/src/components/document-bar/index.js
Comment (jeryj) [reply packages/editor/src/components/document-bar/index.js:200]: I don't think this impacts it one way or another. The main issue with the accessibility of this area is the semantics and inconsistent naming: https://github.com/WordPress/gutenberg/issues/51394. 

1. Buttons should have consistent descriptive names, so this button needs to be visually and consistently named as "Command Palette" or something similar. 
2. The `<h1>` is nested within the `<button>` element. `<button>` elements often strip the semantics of its contents, so the `<h1>` would never be exposed. It is as if there is no heading on the page at all.

So, those core issues do need to be fixed, but this PR doesn't seem to impact the core issues either way.

File: packages/editor/src/components/document-bar/index.js
Comment (creativecoder) [reply packages/editor/src/components/document-bar/index.js:58]: Done in https://github.com/WordPress/gutenberg/pull/65170/commits/7997eb3e80563e44e034f9d1d946a4bfab7d7ba2

**Comment by creativecoder** [comment]
I've pushed some updates that

- Adjust the `IconType` import
- Use darker text color for required contrast on post type label
- Adds `font-weight: 400` css to the document bar h1
- Adjusts the overflow for long titles so it doesn't truncate the post type label

The responsive styles still need some work, though they are currently not much worse than they are on trunk:

image

image

The markup here is not only an accessibility problem, it makes responsive styling more difficult than it needs to be.

File: packages/editor/src/components/document-bar/index.js
Comment (creativecoder) [reply packages/editor/src/components/document-bar/index.js]: Thanks! Adjusted here: https://github.com/WordPress/gutenberg/pull/65170/commits/7997eb3e80563e44e034f9d1d946a4bfab7d7ba2

**Comment by andrewserong** [comment]
Nice work @creativecoder! I've pushed a couple of tiny commits to get the e2e tests passing since we're agreed on removing the `aria-label`.

I did a bit more hacking, and I think I've come up with a way to improve the responsive behaviour for now, if we're willing to accept hiding the post type label when the Back button is present, at an in-between viewport size. It's a little awkward adding the media query, but seems to work okay if we also move the `max-width` up again to the `editor-document-bar__title` and use `flex` values to control the sizing of the title relative to the label. Here's a potential diff:

```diff
diff --git a/packages/editor/src/components/document-bar/style.scss b/packages/editor/src/components/document-bar/style.scss
index bd85ab0db05..2308554d473 100644
--- a/packages/editor/src/components/document-bar/style.scss
+++ b/packages/editor/src/components/document-bar/style.scss
@@ -25,6 +25,14 @@
 			background: $gray-200;
 		}
 	}
+
+	&.has-back-button {
+		@media screen and (min-width: #{ ($break-medium) }) and (max-width: $break-large) {
+			.editor-document-bar__post-type-label {
+				display: none;
+			}
+		}
+	}
 }
 
 .editor-document-bar__command {
@@ -37,6 +45,7 @@
 	overflow: hidden;
 	color: $gray-900;
 	margin: 0 auto;
+	max-width: 70%;
 
 	// Offset the layout based on the width of the ⌘K label. This ensures the title is centrally aligned.
 	@include break-medium() {
@@ -55,7 +64,7 @@
 
 .editor-document-bar__post-title {
 	color: currentColor;
-	max-width: 64%;
+	flex: 2;
 	overflow: hidden;
 	text-overflow: ellipsis;
 
@@ -65,6 +74,7 @@
 }
 
 .editor-document-bar__post-type-label {
+	flex: 0;
 	color: $gray-800;
 	padding-left: $grid-unit-05;
 }
```

The result would be this:

https://github.com/user-attachments/assets/6d897883-74d9-42f6-8aca-d7297b90249b

(Happy to commit to this branch, but I didn't want to make any changes in case there were particular decisions you'd wanted to hold onto in the current shape of the CSS 🙂)

**Comment by creativecoder** [comment]
@andrewserong That approach for responsive styles seems like a reasonable compromise. Please feel free to commit to this branch! I appreciate the help--this has already taken a lot more time than I was expecting.

**Comment by creativecoder** [comment]
> Should we display the area here rather than the generic "Template part"?

@jameskoster I'm open to it, but while trying it out, it seems redundant for the themes I've tested. A couple examples:

image
image

compared to using the post type

image

Not a big sample size, but the block themes I see put "Header" or "Footer" in the names of those template parts. For other template part areas, I'm not sure that the "General" label adds useful information compared to just "Template Part" (e.g. "Sidebar · General Area" vs  "Sidebar · Template Part")

Do you have any other suggestions for the wording and format of the template part area label that would make it more informative?

**Review by andrewserong — APPROVED** [review]
Thanks @creativecoder! I've pushed the fix for that in-between breakpoint and using `flex` rules where appropriate. After re-testing this is working nicely for me now, I've tested a bunch of different screens in Chrome, FF, Edge, and Safari, and it's holding up nicely for me.

I'll give it the tentative approval, but might be good to get a +1 from designers just to make sure it's looking good to land as an iteration. And we can always keep tweaking in follow-ups.

(Also, I noticed a couple if intermittent e2e test failures, but I'm pretty sure they're unrelated to this PR, 🤞 they pass from running them again)

LGTM! 🎉

**Comment by andrewserong** [comment]
Re: failing e2e tests — I believe it's due to a change in `wordpress-develop`. Posted about it here: https://wordpress.slack.com/archives/C02QB2JS7/p1726210812176909

**Comment by jameskoster** [comment]
> Do you have any other suggestions for the wording and format of the template part area label that would make it more informative?

Not off the top of my head. I agree that `[ Header • Header ]` isn't very helpful. Probably a detail to revisit later.

**Comment by andrewserong** [comment]
Alrighty, I've rebased this PR and the tests are passing now. @WordPress/gutenberg-design any objections if we land this PR in its current form, or were there any design tweaks you'd like changed before merge?

**Comment by creativecoder** [comment]
@andrewserong There's one more thing with the responsive styles I'd like to fix: the post type shows instead of the title at very small screen widths:

image

I think this should be justified such that the post type clips rather than the title (or maybe the post type just doesn't show until there's enough room for it.)

Other than that, this is looking good to me. Will tackle issue tomorrow, unless you beat me too it.

**Comment by andrewserong** [comment]
> I think this should be justified such that the post type clips rather than the title (or maybe the post type just doesn't show until there's enough room for it.)
> Other than that, this is looking good to me. Will tackle issue tomorrow, unless you beat me too it.

Oh, good catch! Maybe we can just hide the post type on very small screens, similar to what we've done with that in-between break point. I've had a go at that in https://github.com/WordPress/gutenberg/pull/65170/commits/6f9ae9a1edd00ac6821e7befb9291c802c1cd0ba, but feel free to revert if there's a better way to go about it (or of course land this if it's looking good 😄).

Here's how it's looking to me now:

https://github.com/user-attachments/assets/dd05ce50-bf6c-4235-aa4b-ebcb885c6753

**Comment by andrewserong** [comment]
I think this is in a good place now, and there haven't been any objections so far, so I'll merge this in. Happy to help on any follow-ups if need be!

**Comment by creativecoder** [comment]
Thanks for finishing this one up @andrewserong , much appreciated!

**Pull request merged**
